### ISSUE TITLE
Cop idea: numbered variable names (user_1 vs user1)

### ISSUE THREAD

**Issue by mockdeep** [bbatsov/rubocop#3167]
We've got a lot of places in our code where we use a couple of records, something like:

```ruby
user_1 = blah
user_2 = bloo
```

Sometimes people number them instead, `user1` and `user2`. Would be nice to have a rule to enforce consistent numbering of variable names, so always with `_` or always without.

**Comment by Drenmi** [comment]
Would it also be useful to have a configuration option to disallow numbered variables? I can see how some teams might want this to encourage better design (i.e. using a data structure like `Array`) and naming conventions.

**Comment by mockdeep** [comment]
I can't say that it would be something that we would use. We do a lot of testing around things like one-off scripts and database scopes to ensure that only specific records are returned or modified. So something like:

```ruby
user_1 = create(:user, ...)
user_2 = create(:user, ...)
expect do
  SomeThing.call
end.to change { user_1.reload.some_attribute }
  .and not_change { user_2.reload.some_attribute }
```

**Comment by mikegee** [comment]
I agree with both of you. I think this new cop would have some value and disallowing numbered variables might be the best default config.

**Comment by Drenmi** [comment]
There is, of course, a small chance for false positives. Consider, for example, a method call `base64` within a class, which would get flagged.

**Comment by mikegee** [comment]
What if the cop specifically looked for more than one local variable with incrementing suffix numbers? Perhaps even starting at 0 or 1?

I think that would catch the problem and not miss any violations.

**Comment by mockdeep** [comment]
It almost sounds like there are multiple rules being discussed here:

1) (my suggestion) variable names with a number should all be consistently `name_1` or `name1`
2) variable names should not be named with a number
3) variable names with numbers should be sequential starting from 1 (or 0)
4) (another I just thought of) variable names shouldn't have a number if there's only one of them

**Comment by Drenmi** [comment]
@mockdeep: These would almost certainly go in the same cop (with configuration options), which is why we're discussing them all in the same thread. 😀 Checking adjacent variables is a bit more involved, but might be possible if the rules are well defined.
